_posts/2023-04-05-deep-learning-scientific-computing-R-torch/deep-learning-scientific-computing-R-torch.Rmd

```diff
@@ -0,0 +1,66 @@
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+First things first: Where can you get it? As of today, you can download the [e-book](https://www.taylorfrancis.com/books/mono/10.1201/9781003275923/deep-learning-scientific-computing-torch-sigrid-keydana) or order a [print copy](https://www.routledge.com/Deep-Learning-and-Scientific-Computing-with-R-torch/Keydana/p/book/9781032231396) from the publisher, CRC Press; the free on-line edition is [here](https://skeydan.github.io/Deep-Learning-and-Scientific-Computing-with-R-torch/). There is, to my knowledge, no problem to perusing the on-line version -- besides one: It doesn't have the squirrel that's on the book cover.
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "A red squirrel on a tree, looking attentively."}
+knitr::include_graphics("images/squirrel.png")
+```
+
+So if you're a lover of amazing creatures ...
+
+### What's in the book?
+
+Deep Learning and Scientific Computing with R `torch` has three parts.
+
+The first covers the indispensible basics: tensors, and how to manipulate them; automatic differentiation, the *sine qua non* of deep learning; optimization, the strategy that drives most of what we call artificial intelligence; and neural-network modules, `torch's` way of encapsulating algorithmic flow. The focus is on understanding the concepts, on how things "work" -- that's why we do things like code a neural network from scratch, something you'll probably never do in later use.
+
+Foundations laid, part two -- considerably more sizeable -- dives into deep-learning applications. It is here that the ecosystem surrounding core `torch` enters the spotlight. First, we see how `luz` renders unnecessary many manual programming tasks related to training workflow, performance evaluation, and prediction. Making use of the wrappers and instrumentation facilities it provides, we next learn about two aspects of deep learning no real-world application can afford to neglect: How to make models generalize to unseen data, and how to accelerate training. Techniques we introduce keep re-appearing throughout the use cases we then look at: image classification and segmentation, regression on tabular data, time-series forecasting, and classifying speech utterances. It's in working with images and sound that essential ecosystem libraries, namely, `torchvision` and `torchaudio`, make their appearance, to be used for domain-dependent functionality.
+
+In part three, we move beyond deep learning, and explore how `torch` can figure in general mathematical or scientific applications. Prominent topics are regression using matrix decompositions, the Discrete Fourier Transform, and the Wavelet Transform. Again, the goal is to understand how it works, and why it works like that (at least, to some degree). That's why, here just like in part one, we code algorithms from scratch, before introducing the speed-optimized `torch` equivalents.
```

Comment:
> Again, the goal is to understand how it works, and why it works like that

Is "it" here referring to torch? I might reorder these sentences, or combine them. Something like:

Just like in part one, we code algorithms from scratch before introducing the speed-optimized `torch` equivalents. This allows us to understand how and why `torch` works the way it does.

Or you could intermingle parts of the second with the first, e.g.:
With the goal of understanding how and why `torch` works the way it does, we begin by coding the algorithms from scratch. Then, we move on to the speed-optimized `torch` equivalents.

As always, your call.

Reply:
Oh thanks, your alternative suggestions make it clear to me that my formulation is way too fuzzy ... I really should say more clearly what I mean, which is  not why torch works the way it does, but the underlying ideas ...
I'll think about how I can express that better.

Reply:
what do you think of this - rather big a change, but really, just being honest ...

>The primary goal here is to understand the underlying ideas, and why they are so important. 

I think that if you consider the complete context (talking very concretely about coding)

> The primary goal here is to understand the underlying ideas, and why they are so important. That's why, here just like in part one, we code algorithms from scratch, before introducing the speed-optimized torch equivalents.

it might be fine?

Reply:
Yep—these both work. I think the confusion arose from the fact that the original sentence had "it" and not "they," since there were multiple algorithms. So, either way, the new language makes it much clearer.

Reply:
glad to hear you think so!